### ISSUE TITLE
Support for Boolean or Integer only variables?

### ISSUE THREAD

**Issue by unography** [nucleic/kiwi#80]
Hi,

Is there a way currently to add only Boolean or Integer only Variables?

I'm currently trying to add a constraint for 2 rectangles to make sure that they don't overlap. For this, I have a solution using boolean variables, to check if a rectangle is to the left and above of each other. But I am unable to add a boolean variable.

How do I solve this using kiwi?

**Comment by sccolbert** [comment]
Just setup constraints using the position of the edges. For example, to ensure that rectange `B` remains to the right of rectangle `A`, do something like this:

```
(A.left + A.width) <= B.left
```

**Comment by unography** [comment]
@sccolbert so I don't what to explicitly specify that `A` should always be left of `B`, either way is fine for me, as long as it doesn't overlap.

Like an `OR` condition.

How do I add that?

**Comment by sccolbert** [comment]
OR conditions are difficult/impossible to model, because by definition there is more than one solution to the problem. Kiwi is not a good solver for that type of problem.

**Comment by unography** [comment]
Oh, okay.
Also, is it possible to take absolute and square values of Variables?

Like I want to do operations like - 

```
val = abs(A.mid_x - B.mid_x)
```
and 
```
val = (A.mid_x - B.mid_x)^2 + (A.mid_y - B.mid_y)^2
```

**Comment by sccolbert** [comment]
Nope. Kiwi is a linear programming solver. All constraints must be linear and continuous.

**Comment by unography** [comment]
Okay, thanks!

**Comment by Qix-** [comment]
Sorry for gravedigging a bit. Just wanted to share my findings.

Knowing full-well it was probably a waste of time, I wanted to see what benefits, if any, there would be by allowing Kiwi to be created using integral types instead of `double`.

My attempt can be seen at https://github.com/Qix-/kiwi/tree/int-solver.

This changed `Solver`, `Term`, `Constraint`, `Variable` and `Expression`s to `BasicSolver`, `BasicTerm`, etc. and templated them with a `TValue` type. It then created aliases via e.g. `using Variable = BasicVariable<double>;` and `using IntVariable = BasicVariable<long long>;`, etc.

What I found (and what is probably obvious to those who are already familiar with Cassowary) is that the internal algorithm relies heavily on linear algebra and the use of coefficients, which cannot be expressed neatly as integer values. A lot of normalization happens, which means values expressed in sub-integral domains, which means that even though the API successfully enforces integral types, **the internal representation _must_ remain `double` for the algorithm to work at all.**

The initial attempt was to change all instances of `double` to `TValue` save for weights, which of course did not work (as I expected would probably be the case). Moving backward to revert those to `double` where needed in order to get the algorithm working at all resulted in reverting all of the coefficients handling and the algorithm itself back to `double` types.

The two outcomes of this are thus:

- There is no perceivable difference in output aside from simply truncating the `variable.value()` result.
- Adding integral API typing only adds extra floating point &lt;-&gt; integer conversions, which are _slow_ and will inevitably slow down the overall execution as opposed to simply converting values to `double` once and from `double` once (after updating the solver).

There is absolutely zero functional benefit to adding integer-only variables to Kiwi, definitively.
 
I wrote this so that it might be referred to in the future if the question ever came up again, just as a reference. My particular use-case was that the 'canvas' onto which elements were being laid out required that the positions be integral. I wanted to make sure that specifying an element being below another element didn't result in `0.5` values, which would truncate to `0`.

However, this isn't possible given correct constraints in Kiwi, namely because there is no concept of plain `>` and `<` comparisons - only `>=` and `<=`, which means that the offsets need to be at least `1` anyway - and further, the truncation in the case of integer-only variables would still happen *after* the algorithm runs anyway, so the 'problem' couldn't be fixed by ditching floating point altogether.

Again, this probably seems obvious to the maintainers, but it was a fun exercise to see what was possible and to see how the internals worked a bit more. I can't claim to understand the Cassowary algorithm in its entirety but at least I understand how Kiwi is structured a lot better now.
